simpl/incubator/rest.py

```diff
@@ -117,62 +117,27 @@ def validate(val):
     return validate
 
 
-def schema(body_schema=None, body_required=False, query_schema=None,
-           content_types=None, default_body=None):
-    """Decorator to parse and validate API body and query string.
+def query(schema=None):
+    """Decorator to parse and validate API query string.
 
     This decorator allows one to define the entire 'schema' for an API
     endpoint.
 
-    :keyword body_schema:
-        Callable that accepts raw data and returns the coerced (or unchanged)
-        body content if it is valid. Otherwise, an error should be raised.
-
-    :keyword body_required:
-        `True` if some body content is required by the request. Defaults to
-        `False`.
-
     :keyword query_schema:
         Callable that accepts raw data and returns the coerced (or unchanged)
         query string content if it is valid. Otherwise, an error should be
         raised.
-
-    :keyword content_types:
-        List of allowed contents types for request body contents. Defaults to
-        `['application/json']`.
-
-    :keyword default_body:
-        Default body value to pass to the endpoint handler if `body_required`
-        is `True` but no body was given. This can be useful for specifying
-        complex request body defaults.
     """
-    if not content_types:
-        content_types = ['application/json']
-    if not all('json' in t for t in content_types):
-        raise NotImplementedError("Only 'json' body supported.")
-
     def deco(func):
         """Return a decorated callable."""
         def wrapped(*args, **kwargs):
-            """Validate/coerce request body and parameters."""
+            """Validate/coerce request query parameters."""
             try:
-                # validate the request body per the schema (if applicable):
-                body = bottle.request.json
-                if body is None:
-                    body = default_body
-                if body_required and not body:
-                    bottle.abort(400, 'Call body cannot be empty')
-                if body_schema:
-                    try:
-                        body = body_schema(body)
-                    except volup.MultipleInvalid as exc:
-                        raise MultiValidationError(exc.errors)
-
                 # validate the query string per the schema (if application):
                 query = bottle.request.query.dict
                 if query_schema is not None:
                     try:
-                        query = query_schema(query)
+                        query = schema(query)
```

Comment:
I'm still looking for the best way to integrate with pagination and `process_params`.

Our pagination and `process_params` code is very opinionated. So keeping it out of `rest.query` would make `rest.query` more flexible, but we've been approaching `simpl` as a very opinionated library/framework, so I would say we should actually couple `rest.query` with `process_params` more tightly first and then iterate on them together. Otherwise we're introducing two different ways to handle query params, which is not opinionated.

A good first place to start would be to have `rest.query` be effectively a decorator version of `process_params` that also accepts a custom schema.

```python
@bottle.get('/v1/widgets/')
# same signature as `process_params` with the addition of a schema kwarg for
# validating any extra data
@rest.query(filter_fields=['status'], schema={'cache': str})
@rest.paginated(resource_name='widget'})
def list_widgets(status=None, limit=0, offset=0, query=None):
    """List widgets in pages filtered by optional status[es]."""
    return query_widgets(limit=limit, offset=offset, **query)
```
In the above example, all the output of `process_params` would be considered valid and only any remaining query params would be validated by the schema so that we wouldn't have to define status, limit, and offset again in query.

If `query` has all of `limit, status, offset` and `cache`, then `query` is a good name for it. But if it only has `cache` then maybe we rename it to `extra_params` which would look like this:


```python
@bottle.get('/v1/widgets/')
@rest.query(filter_fields=['status'], schema={'cache': str})
@rest.paginated(resource_name='widget'})
def list_widgets(status=None, limit=0, offset=0, extra_prams=None):
    """List widgets in pages filtered by optional status[es]."""
    assert extra_params and extra_params.keys() == ['cache']
    return query_widgets(status=status, limit=limit, offset=offset, **extra_params)
```

So, practically, I think the following would accomplish the integration:
1 - `process_params` should be extended to support a custom schema argument that is used to validate any extra params that are not part of filter_fields or standard_params.
2 - `rest.query` should have pretty much the same signature as process_params
3 - the line above should call process_params.

I think there will be opportunities to refactor while doing that and `rest.query` would become a very small wrapper for `process_params`.